File: tests/test_doom_wad.py
import struct
import tempfile
import unittest
from pathlib import Path

from doom_wad import WadFile, Palette, build_wad_bytes


class WadFileExampleTest(unittest.TestCase):
    """examples/doom/wad.md 의 # Examples 를 그대로 옮긴 테스트."""

    def setUp(self):
        self.tmpdir = tempfile.TemporaryDirectory()
        self.path = Path(self.tmpdir.name) / "doom1.wad"
        data = build_wad_bytes([
            ("E1M1", b"map-data-goes-here"),
            ("PLAYPAL", bytes(range(256)) * 3),
        ])
        self.path.write_bytes(data)

    def tearDown(self):
        self.tmpdir.cleanup()

    def test_find_and_read_known_lump(self):
        wad = WadFile(str(self.path))
        idx = wad.findLump("E1M1")
        data = wad.readLump(idx)
        self.assertNotEqual(idx, -1)
        self.assertEqual(len(data), wad.lumps[idx].size)
        self.assertEqual(data, b"map-data-goes-here")


class WadFileRequiredTest(unittest.TestCase):
    """wad.md의 test_required Doom.WadFile 블록을 그대로 옮긴 테스트."""

    def test_read_lump_length_matches_size(self):
        data = build_wad_bytes([("A", b"1234567"), ("B", b"xy")])
        f = tempfile.NamedTemporaryFile(suffix=".wad", delete=False)
        f.write(data)
        f.close()
        wad = WadFile(f.name)
        for i, lump in enumerate(wad.lumps):
            self.assertEqual(len(wad.readLump(i)), lump.size)

    def test_missing_name_returns_minus_one(self):
        data = build_wad_bytes([("A", b"x")])
        f = tempfile.NamedTemporaryFile(suffix=".wad", delete=False)
        f.write(data)
        f.close()
        wad = WadFile(f.name)
        self.assertEqual(wad.findLump("NOPE"), -1)

    def test_duplicate_name_returns_latest(self):
        data = build_wad_bytes([("DUP", b"old"), ("DUP", b"newer")])
        f = tempfile.NamedTemporaryFile(suffix=".wad", delete=False)
        f.write(data)
        f.close()
        wad = WadFile(f.name)
        idx = wad.findLump("DUP")
        self.assertEqual(wad.readLump(idx), b"newer")

    def test_case_insensitive_lookup(self):
        data = build_wad_bytes([("E1M1", b"x")])
        f = tempfile.NamedTemporaryFile(suffix=".wad", delete=False)
        f.write(data)
        f.close()
        wad = WadFile(f.name)
        self.assertEqual(wad.findLump("e1m1"), 0)


class WadFileMagicTest(unittest.TestCase):
    def test_rejects_non_wad_file(self):
        f = tempfile.NamedTemporaryFile(suffix=".wad", delete=False)
        f.write(b"NOTAWADFILE" + b"\x00" * 20)
        f.close()
        with self.assertRaises(ValueError):
            WadFile(f.name)


class PaletteTest(unittest.TestCase):
    def test_color_at_reads_rgb_triplet(self):
        playpal = bytes([10, 20, 30, 40, 50, 60]) + bytes(768 - 6)
        pal = Palette(playpal)
        self.assertEqual(pal.colorAt(0), (10, 20, 30))
        self.assertEqual(pal.colorAt(1), (40, 50, 60))


if __name__ == "__main__":
    unittest.main()

File: tests/cpp/test_doom_wad.cpp
// examples/doom/wad.md 검증 (C++17).
#include "doom_wad.hpp"
#include "minitest.hpp"
#include <cstdio>
#include <fstream>

using namespace Doom;

static std::string writeTempWad(const std::vector<uint8_t>& bytes, const char* suffix) {
    std::string path = std::string("tmp_wad_test") + suffix + ".wad";
    std::ofstream f(path, std::ios::binary);
    f.write(reinterpret_cast<const char*>(bytes.data()), bytes.size());
    f.close();
    return path;
}

TEST(wad_find_and_read_known_lump) {
    auto bytes = buildWadBytes({{"E1M1", {'m', 'a', 'p', '-', 'd', 'a', 't', 'a'}}});
    auto path = writeTempWad(bytes, "_1");
    WadFile wad(path);
    int idx = wad.findLump("E1M1");
    CHECK(idx != -1);
    auto data = wad.readLump(idx);
    CHECK_EQ(static_cast<int>(data.size()), wad.lumps[idx].size);
    std::remove(path.c_str());
}

TEST(wad_missing_name_returns_minus_one) {
    auto bytes = buildWadBytes({{"A", {1}}});
    auto path = writeTempWad(bytes, "_2");
    WadFile wad(path);
    CHECK_EQ(wad.findLump("NOPE"), -1);
    std::remove(path.c_str());
}

TEST(wad_duplicate_name_returns_latest) {
    auto bytes = buildWadBytes({{"DUP", {1, 2, 3}}, {"DUP", {9, 9}}});
    auto path = writeTempWad(bytes, "_3");
    WadFile wad(path);
    int idx = wad.findLump("DUP");
    auto data = wad.readLump(idx);
    CHECK_EQ(data.size(), 2u);
    CHECK_EQ(data[0], 9);
    std::remove(path.c_str());
}

TEST(wad_case_insensitive_lookup) {
    auto bytes = buildWadBytes({{"E1M1", {1}}});
    auto path = writeTempWad(bytes, "_4");
    WadFile wad(path);
    CHECK_EQ(wad.findLump("e1m1"), 0);
    std::remove(path.c_str());
}

TEST(wad_rejects_non_wad_file) {
    std::vector<uint8_t> junk(30, 0);
    junk[0] = 'N'; junk[1] = 'O'; junk[2] = 'P'; junk[3] = 'E';
    auto path = writeTempWad(junk, "_5");
    bool threw = false;
    try {
        WadFile wad(path);
    } catch (const std::exception&) {
        threw = true;
    }
    CHECK(threw);
    std::remove(path.c_str());
}

TEST(palette_color_at_reads_rgb_triplet) {
    std::vector<uint8_t> playpal(768, 0);
    playpal[0] = 10; playpal[1] = 20; playpal[2] = 30;
    playpal[3] = 40; playpal[4] = 50; playpal[5] = 60;
    Palette pal(playpal);
    auto c0 = pal.colorAt(0);
    CHECK_EQ(c0[0], 10); CHECK_EQ(c0[1], 20); CHECK_EQ(c0[2], 30);
    auto c1 = pal.colorAt(1);
    CHECK_EQ(c1[0], 40); CHECK_EQ(c1[1], 50); CHECK_EQ(c1[2], 60);
}

TEST_MAIN()
